qiskit_experiments/data_processing/data_processor.py

```diff
@@ -172,35 +190,80 @@ def _data_extraction(self, data: Union[Dict, List[Dict]]) -> List:
         Args:
             data: A list of such dicts where the data is contained under the key self._input_key.
 
-        Returns:
-            The data formatted in such a way that it is ready to be processed by the nodes.
+        Yields:
+            A tuple of numpy array object representing a data and error.
 
         Raises:
             DataProcessorError:
                 - If the input datum is not a list or a dict.
-                - If the data processor received a single datum but requires all the data to
-                  process it properly.
                 - If the input key of the data processor is not contained in the data.
         """
         if isinstance(data, dict):
             data = [data]
 
-        try:
-            data_ = [_datum[self._input_key] for _datum in iter(data)]
-        except KeyError as error:
-            raise DataProcessorError(
-                f"The input key {self._input_key} was not found in the input datum."
-            ) from error
-        except TypeError as error:
-            raise DataProcessorError(
-                f"{self.__class__.__name__} only extracts data from "
-                f"lists or dicts, received {type(data)}."
-            ) from error
-
-        return data_
+        for datum in data:
+            try:
+                target = datum[self._input_key]
+
+                # returns data and initial error
+                if isinstance(target, dict):
+                    # likely level2 data, forcibly convert into array
+                    yield np.asarray([target], dtype=object), np.asarray([np.nan], dtype=float)
+                else:
+                    # level1 or below
+                    nominal_arr = np.asarray(target, dtype=float)
+                    stdev_arr = np.full_like(target, np.nan, dtype=float)
+
+                    yield nominal_arr, stdev_arr
+
+            except KeyError as error:
+                raise DataProcessorError(
+                    f"The input key {self._input_key} was not found in the input datum."
+                ) from error
+            except TypeError as error:
+                raise DataProcessorError(
+                    f"{self.__class__.__name__} only extracts data from "
+                    f"lists or dicts, received {type(data)}."
+                ) from error
 
     def __repr__(self):
         """String representation of data processors."""
         names = ", ".join(node.__class__.__name__ for node in self._nodes)
 
         return f"{self.__class__.__name__}(input_key={self._input_key}, nodes=[{names}])"
+
+
+def execute_pipeline(gen_datum: Iterator) -> Tuple[np.ndarray, np.ndarray]:
+    """Execute processing pipeline and return processed data array.
+
+    Args:
+        gen_datum: A generator to sequentially return datum.
+
+    Returns:
+        A tuple of nominal values and standard errors.
+    """
+    out_values, out_errors = list(zip(*gen_datum))
```

Comment:
Why is this getting an Iterator when we are immediately consuming it as a list? Would it not be better to give it a list directly?

Reply:
The purpose of this function is to run `list(zip(*gen_datum))` (along with some formatting). A generator is a generator object and you cannot access each element until it is executed like this. Once we integrate error into value with [uncertainties](https://pypi.org/project/uncertainties/) this can be `return np.fromiter(gen_datum, dtype=object)`.